File: src/orders/createOrder.ts
import { DynamoDB } from 'aws-sdk';
import { Category } from 'aws-sdk/clients/signer';

import { SqsEvent } from '../sqs';
import { Order, OrderProduct, OrderStatus } from './forms';

const ddb = new DynamoDB.DocumentClient({ region: "eu-central-1" })

export async function handler(event: SqsEvent) {
  const [{ body }] = event.Records
  const order = JSON.parse(body) as Order

  const sumAry: number[] = await Promise.all(order.products.map(p => calcPriceSum(p)))
  const sum: number = sumAry.reduce((a, p) => a + p, 0)

  const params = {
    TableName: process.env.ORDERS_TABLE!,
    Item: {
      userId: order.userId,
      orderNo: order.orderNo,
      name: order.name,
      email: order.email,
      street: order.street,
      housenr: order.housenr,
      city: order.city,
      postcode: order.postcode,
      sum: sum,
      products: order.products,
      orderStatus: OrderStatus.Confirmed
    }
  }

  // Validate order (check stock etc.)
  const promAry = await Promise.all(order.products.map(p => checkStock(p.category, p.productId, p.amount)))
  const allAvailable = promAry.every(e => e === true)

  if (allAvailable) {
    try {
      await ddb.put(params).promise()
      //update stock
      await updateStock(order.products)
      //send order confirming e-mail to customer
      await emptyBasket(order.userId)
    } catch (err) {
      console.log(err)
      //send apologizing e-mail to customer (due to internal failure)
    }
  } else {
    console.log('Out of Stock')
    //change status from confirmed to 
    params.Item.orderStatus = OrderStatus.OutOfStock
    await ddb.put(params).promise()
    // send apologizing e-mail to customer (out of stock)
  }
}

export async function calcPriceSum(p: OrderProduct): Promise<number> {
  const params = {
    TableName: process.env.PRODUCTS_TABLE!,
    Key: {
      category: p.category,
      productId: p.productId
    },
  }
  const data = await ddb.get(params).promise()
  return data.Item.price * p.amount
}

export async function checkStock(category: Category, productId: string, amount: number): Promise<boolean> {
  const params = {
    TableName: process.env.PRODUCTS_TABLE!,
    Key: {
      category: category,
      productId: productId
    },
  }
  const data = await ddb.get(params).promise()
  //if((data.Item.stock - amount) <= 0){
  //alert: reorder neccecary
  //}
  if (data.Item.stock >= amount) {
    return true
  } else {
    return false
  }
}

export async function emptyBasket(userId: string): Promise<any> {
  const params = {
    TableName: process.env.BASKET_TABLE!,
    Key: {
      userId: userId
    },
    UpdateExpression: "set products=:p",
    ExpressionAttributeValues: {
      ":p": [],
    },
    ReturnValues: "ALL_NEW"
  }
  return ddb.update(params).promise()
}

export async function updateStock(products: OrderProduct[]): Promise<any> {
  return Promise.all(products.map(p => {
    const params = {
      TableName: process.env.PRODUCTS_TABLE!,
      Key: {
        category: p.category,
        productId: p.productId
      },
      UpdateExpression: "set stock=stock-:a",
      ExpressionAttributeValues: {
        ":a": p.amount,

      },
      ReturnValues: "UPDATED_NEW"
    }
    return ddb.update(params).promise()
  }))
}

File: dynamo_go/lambda/orders/createOrder/createOrder.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
	"os"
	"webshop/internal/forms"
)

func Handler(event events.SNSEvent) error {
	ddb := dynamodb.New(session.New())

	var orderRequest forms.OrderRequest
	if err := json.Unmarshal([]byte(event.Records[0].SNS.Message), &orderRequest); err != nil {
		return err
	}

	item, err := dynamodbattribute.MarshalMap(orderRequest)
	if err != nil {
		return err
	}

	//calculate sum of orderProduct prices
	sum := 0.0
	for _, p := range orderRequest.Products {
		res, err := calcSum(p, ddb)
		if err != nil {
			return err
		}
		sum += res
	}

	sumAttr, err := dynamodbattribute.Marshal(sum)
	if err != nil {
		return err
	}
	item["sum"] = sumAttr

	//check if all products available
	var allAvailable bool
	for _, p := range orderRequest.Products {
		res, err := checkStock(p, ddb)
		if err != nil {
			return err
		}
		allAvailable = res
	}

	if allAvailable {
		item["orderStatus"] = &dynamodb.AttributeValue{S: aws.String("Confirmed")}
		//update Stock
		err := updateStock(orderRequest.Products, ddb)
		if err != nil {
			return err
		}
		//empty Basket
		e := emptyBasket(orderRequest.UserId, ddb)
		if e != nil {
			return e
		}
	} else {
		item["orderStatus"] = &dynamodb.AttributeValue{S: aws.String("OutOfStock")}
	}

	_, err = ddb.PutItem(&dynamodb.PutItemInput{
		TableName: aws.String(os.Getenv("ORDERS_TABLE")), //aws.String macht wert zu einem  pointer
		Item:      item,
	})

	if err != nil {
		return err
	}

	return nil
}

func emptyBasket(id *string, ddb *dynamodb.DynamoDB) error {
	_, err := ddb.UpdateItem(&dynamodb.UpdateItemInput{
		TableName: aws.String(os.Getenv("BASKET_TABLE")), //aws.String macht wert zu einem  pointer
		Key: map[string]*dynamodb.AttributeValue{
			"userId": {S: id},
		},
		UpdateExpression: aws.String("set products=:p"),
		ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
			":p": {L: []*dynamodb.AttributeValue{}},
		},
		ReturnValues: aws.String("ALL_NEW"),
	})
	if err != nil {
		return err
	}
	return nil
}

func updateStock(products []*forms.OrderProduct, ddb *dynamodb.DynamoDB) error {
	for _, p := range products {
		_, err := ddb.UpdateItem(&dynamodb.UpdateItemInput{
			TableName: aws.String(os.Getenv("PRODUCTS_TABLE")), //aws.String macht wert zu einem  pointer
			Key: map[string]*dynamodb.AttributeValue{
				"category":  {S: aws.String(p.Category)},
				"productId": {S: aws.String(p.ProductId)},
			},
			UpdateExpression: aws.String("set stock=stock-:a"),
			ExpressionAttributeValues: map[string]*dynamodb.AttributeValue{
				":a": {N: aws.String(fmt.Sprint(p.Amount))},
			},
			ReturnValues: aws.String("UPDATED_NEW"),
		})
		if err != nil {
			return err
		}
	}
	return nil
}

func checkStock(p *forms.OrderProduct, ddb *dynamodb.DynamoDB) (bool, error) {
	var product forms.Product
	out, err := ddb.GetItem(&dynamodb.GetItemInput{
		TableName: aws.String(os.Getenv("PRODUCTS_TABLE")), //aws.String macht wert zu einem  pointer
		Key: map[string]*dynamodb.AttributeValue{
			"category":  {S: aws.String(p.Category)},
			"productId": {S: aws.String(p.ProductId)},
		},
	})
	if err != nil {
		return false, err
	}
	if err := dynamodbattribute.UnmarshalMap(out.Item, &product); err != nil {
		return false, err
	}
	return product.Stock >= p.Amount, err
}

func calcSum(p *forms.OrderProduct, ddb *dynamodb.DynamoDB) (float64, error) {
	var product forms.Product
	out, err := ddb.GetItem(&dynamodb.GetItemInput{
		TableName: aws.String(os.Getenv("PRODUCTS_TABLE")), //aws.String macht wert zu einem  pointer
		Key: map[string]*dynamodb.AttributeValue{
			"category":  {S: aws.String(p.Category)},
			"productId": {S: aws.String(p.ProductId)},
		},
	})
	if err != nil {
		return 0.0, err
	}
	if err := dynamodbattribute.UnmarshalMap(out.Item, &product); err != nil {
		return 0.0, err
	}
	return product.Price * float64(p.Amount), err

}

func main() {
	lambda.Start(Handler)
}
